### PR TITLE
fix: don't remove comments between key and value in object-shorthand

### LINKED ISSUE #18441: Bug: object-shorthand auto-fix removes comments
### Environment

Node version: any
npm version: any
Local ESLint version: 9.2.0
Global ESLint version: 9.2.0
Operating System: any

### What parser are you using?

Default (Espree)

### What did you do?

Run autofix on this:

```js
/* eslint object-shorthand: ["error"] */

let b;

const a = {
  b: /* this comment will be auto-removed */ b,
  c: /* this comment will stay */ function () {}
};
```

### What did you expect to happen?

All comments to stay:

```js
/* eslint object-shorthand: ["error"] */

let b;

const a = {
  b: /* this comment will be auto-removed */ b,
  c: /* this comment will stay */ function () {}
};
```

### What actually happened?

First comment got deleted:

```js
/* eslint object-shorthand: ["error"] */

let b;

const a = {
  b,
  c: /* this comment will stay */ function () {}
};
```

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG9iamVjdC1zaG9ydGhhbmQ6IFtcImVycm9yXCJdICovXG5cbmxldCBiO1xuXG5jb25zdCBhID0ge1xuICBiOiAvKiBhc2QgKi8gYiwgICAvKiA8LS0gY2xpY2sgRklYIGhlcmUgKi9cblxuICBjOiAvKiBhc2QgKi8gZnVuY3Rpb24gKCkge30gICAgLyogPC0tIG5vIEZJWCBhdmFpbGFibGUgaGVyZSAqL1xufTtcblxuYTsiLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by KubaJastrz**
Fixes #18441

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Check for comments between the last key token and the value in the property node.

### CODE REVIEW

File: lib/rules/object-shorthand.js [hunk 1]
```diff
@@ -497,6 +497,13 @@ module.exports = {
                         node,
                         messageId: "expectedPropertyShorthand",
                         fix(fixer) {
+                            const lastKeyToken = sourceCode.getLastToken(node.key);
+
+                            // x: /* */ x
+                            if (sourceCode.commentsExistBetween(lastKeyToken, node.value)) {
+                                return null;
+                            }
+
```
Comment (Tanujkanti4441) [inline lib/rules/object-shorthand.js:506]: how about if we give a fix here something like:
```js
var x = {
a /* comment */: a,
b
}
```
can be fixed to
```js
var x = {
/* comment */ a,
b
}
```
or

```js
var x = {
a, /* comment */ 
b
}
```

Comment (KubaJastrz) [reply lib/rules/object-shorthand.js:506]: I think in that case it becomes a feature request, not a bug fix. You can create an issue for this idea.

Whatever the decision is, I'd recommend that both identifiers and function expressions were handled identically (which they aren't right now, hence this PR).

Comment (mdjermanovic) [reply lib/rules/object-shorthand.js:506]: I think it's fine not to provide the fix when there's a comment between the two identifiers and let the author fix this manually as it isn't clear where the new position for the comment should be.

File: tests/lib/rules/object-shorthand.js [hunk 2]
```diff
@@ -516,6 +516,16 @@ ruleTester.run("object-shorthand", rule, {
...
+        {
+            code: "var x = {a: /* comment */ a}",
+            output: null,
+            errors: [PROPERTY_ERROR]
+        },
+        {
+            code: "var x = {a /* comment */: a}",
+            output: null,
+            errors: [PROPERTY_ERROR]
+        },
```
Comment (mdjermanovic) [inline tests/lib/rules/object-shorthand.js:528]: Can you also add tests with quotes to make sure that case is covered.

```suggestion
        },
        {
            code: "var x = {'a': /* comment */ a}",
            output: null,
            errors: [PROPERTY_ERROR]
        },
        {
            code: "var x = {'a' /* comment */: a}",
            output: null,
            errors: [PROPERTY_ERROR]
        },
```

File: lib/rules/object-shorthand.js [hunk 3]
```diff
@@ -497,6 +497,13 @@ module.exports = {
                         node,
                         messageId: "expectedPropertyShorthand",
                         fix(fixer) {
+                            const lastKeyToken = sourceCode.getLastToken(node.key);
+
+                            // x: /* */ x
+                            if (sourceCode.commentsExistBetween(lastKeyToken, node.value)) {
+                                return null;
+                            }
```
Comment (mdjermanovic) [inline lib/rules/object-shorthand.js:505]:
```suggestion

                            // x: /* */ x
                            if (sourceCode.commentsExistBetween(node.key, node.value)) {
                                return null;
                            }
```

We can use `node.key` directly.

File: lib/rules/object-shorthand.js [hunk 4]
```diff
@@ -510,6 +517,13 @@ module.exports = {
                         node,
                         messageId: "expectedPropertyShorthand",
                         fix(fixer) {
+                            const lastKeyToken = sourceCode.getLastToken(node.key);
+
+                            // "x": /* */ x
+                            if (sourceCode.commentsExistBetween(lastKeyToken, node.value)) {
+                                return null;
+                            }
```
Comment (mdjermanovic) [inline lib/rules/object-shorthand.js:525]:
```suggestion

                            // "x": /* */ x
                            if (sourceCode.commentsExistBetween(node.key, node.value)) {
                                return null;
                            }
```

Same here.

File: tests/lib/rules/object-shorthand.js (hunk 2 above)
Comment (KubaJastrz) [reply tests/lib/rules/object-shorthand.js:528]: [6005b24](https://github.com/eslint/eslint/pull/18442/commits/6005b24af15e1dd5e96a7482a866ee1bee3b5170)

File: lib/rules/object-shorthand.js (hunk 3 above)
Comment (KubaJastrz) [reply lib/rules/object-shorthand.js]: [d5c6fe1](https://github.com/eslint/eslint/pull/18442/commits/d5c6fe1f8540f299d010e1256eff97151b995e42)

File: lib/rules/object-shorthand.js (hunk 4 above)
Comment (KubaJastrz) [reply lib/rules/object-shorthand.js]: [d5c6fe1](https://github.com/eslint/eslint/pull/18442/commits/d5c6fe1f8540f299d010e1256eff97151b995e42)

**Review by aladdin-add — DISMISSED** [review]
LGTM, thanks!

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
Just noticed an edge case that wasn't covered (see the new test cases).

File: tests/lib/rules/object-shorthand.js [hunk 5]
```diff
@@ -516,6 +516,26 @@ ruleTester.run("object-shorthand", rule, {
...
```
[the remark this answers is not in the document]
Comment (KubaJastrz) [reply tests/lib/rules/object-shorthand.js:528]: [58fe097](https://github.com/eslint/eslint/pull/18442/commits/58fe09742b2cd1715ed012cb141f5235b490e8b0)

File: tests/lib/rules/object-shorthand.js [hunk 6]
```diff
@@ -516,6 +516,26 @@ ruleTester.run("object-shorthand", rule, {
...
+        {
+            code: "var x = {'a': /* comment */ a}",
+            output: null,
+            errors: [PROPERTY_ERROR]
+        },
+        {
+            code: "var x = {'a' /* comment */: a}",
+            output: null,
+            errors: [PROPERTY_ERROR]
+        },
```
[the remark this answers is not in the document]
Comment (KubaJastrz) [reply tests/lib/rules/object-shorthand.js:548]: [58fe097](https://github.com/eslint/eslint/pull/18442/commits/58fe09742b2cd1715ed012cb141f5235b490e8b0)

File: lib/rules/object-shorthand.js [hunk 7]
```diff
@@ -497,6 +497,12 @@ module.exports = {
                         node,
                         messageId: "expectedPropertyShorthand",
                         fix(fixer) {
+
+                            // x: /* */ x
+                            if (sourceCode.commentsExistBetween(node.key, node.value)) {
```
[the remark this answers is not in the document]
Comment (KubaJastrz) [reply lib/rules/object-shorthand.js]: [58fe097](https://github.com/eslint/eslint/pull/18442/commits/58fe09742b2cd1715ed012cb141f5235b490e8b0)

File: lib/rules/object-shorthand.js [hunk 8]
```diff
@@ -510,6 +516,12 @@ module.exports = {
                         node,
                         messageId: "expectedPropertyShorthand",
                         fix(fixer) {
+
+                            // "x": /* */ x
+                            if (sourceCode.commentsExistBetween(node.key, node.value)) {
```
[the remark this answers is not in the document]
Comment (KubaJastrz) [reply lib/rules/object-shorthand.js]: [58fe097](https://github.com/eslint/eslint/pull/18442/commits/58fe09742b2cd1715ed012cb141f5235b490e8b0)

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**